<a href="https://colab.research.google.com/github/stndb01/Random/blob/master/UTS_PPM_Aldi_Fianda_Putra_195150201111034.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Nama : Aldi Fianda Putra

NIM : 195150201111034

Kelas : PPM TIF - D

UTS PPM

# 1. Apa yang dimaksud dengan Pembelajaran Mesin?

Pembelajaran mesin adalah salah satu cabang atau turunan dari kepintaran buatan atau dikenal dengan AI (Artificial Intelligence). Pembelajaran mesin ini sendiri adalah suatu teknik yang dapat kita terapkan pada mesin ataupun komputer dimana dengan teknik tersebut kita mengajarkan mesin atau komputer untuk melakukan suatu tindakan dengan cara mempelajarinya dari suatu data. Data tersebut dipelajari oleh mesin agar nantinya mesin dapat menentukan tindakan yang akan dilakukan ketika suatu kondisi yang belum pernah terjadi tetapi sudah diprediksi akan terjadi. Sehingga dalam tahapan pembuatan aplikasi pembelajaran mesin dilakukan bukan untuk mencari jawaban, melainkan dari jawaban dan data tersebut dilakukan suatu pembelajaran mesin yang dapat dilakukan dengan bantuan bahasa pemrograman untuk menghasilkan suatu aturan yang dimana akan dijadikan pedoman tindakan bagi mesin kedepannya.

# 2. Apa yang dimaksud dengan Ekstraksi fitur?

Ekstraksi fitur adalah salah satu proses yang dapat dikatakan penting dalam pembelajaran mesin. Ekstraksi fitur ini merupakan kegiatan dimana kita mengambil suatu ciri-ciri yang memiliki sebuah keunikan tersendiri dari suatu data sehingga ciri-ciri yang sama tersebut dijadikan sebuah karakteristik yang akan mewakili objek dengan ciri-ciri yang sama. Sehingga fitur yang diambil haruslah sebuah ciri-ciri dari seuatu objek yang sangat membedakan objek tersebut dengan objek yang lainnya.

# 3. Berikan gambaran dan jelaskan alur dari pembelajaran mesin!

Berikut adalah urutan atau alur dari pembelajaran mesin

#### 1. Mengumpulkan Data

Hal yang pertama kali dan paling penting dilakukan tentunya adalah mengumpulkan data dan juga sensor yang akan digunakan untuk mengambil data tersebut. Data yang diambil ini sendiri adalah nilai dari data yang ada pada setiap objeknya. Biasanya data yang diambil ini dikategorikan nantinya berdasarkan fitur yang akan dipilih. Data yang diambil ini dapat berupa data yang bersifat kualitatif yang mana berarti data yang bersifat bukan angka, yakni seperti string atau kalimat. Dan bisa juga data yang bersifat kuantitatif yakni data yang bersifat angka, data angka ini juga dapat dikategorikan lagi menjadi data diskrit yang mana diperoleh dari pengamatan saja tanpa adanya proses penghitungan atau data kontinu yang mana diperoleh dari proses penghitungan data. 

#### 2. Memilih fitur yang akan diambil

Dari data yang diambil sebelumnya, tentu kita akan memperoleh banyak sekali nilai-nilai yang dapat kita kategorikan agar dapat mempermudah proses penghitungan yang akan dilakukan. Oleh sebab itu kita melakukan ekstraksi fitur untuk mengambil nilai yang mewakili ciri-ciri dari suatu objek. Fitur yang diambil harus fitur yang bersifat unik dimana membedakannya dengan objek yang lainnya.

#### 3. Memilih model

Dari data dan fitur yang telah diperoleh sebelumnya, maka tindakan yang akan kita lakukan berikutnya adalah menentukan model yang akan kita gunakan untuk memproses datanya, model yang digunakan tentu model yang relevan dan dapat bekerja pada data yang digunakan. Contohnya kita menggunakan model algoritma naive bayes gaussian ketika menghitung data numerik dan naive bayes multinomial ketika menghitung data yang bukan numerik seperti string.

#### 4. Melakukan proses training

Tahap ini adalah tahap dimana melatih mesin atau komputer dengan menggunakan model yang telah ditetapkan sebelumnya untuk memperoleh suatu nilai atau tindakan yang ideal bagi mesin untuk digunakan nanti pada kondisi tersebut muncul. Proses training ini sendiri dapat dianalogikan seperti bagaimana manusia itu belajar agar mampu melakukan suatu tindakan sesuai kondisi yang ada.

#### 5. Mengevaluasi hasil dari pembelajaran mesin

# 4. Apa keuntungan dilakukan pemrosesan data awal dan berikan contoh algoritma pemrosesan data awal

Dengan melakukan preprocessing atau pemrosesan data awal kita akan memperoleh beberapa keuntungan yang sangat penting dalam memproses data kita nantinya, yakni sebagai berikut :

#### 1. Membersihkan data dari hal-hal yang mengganggu proses penghitungan 

Pada dataset seringkali kita menemukan data-data yang dapat mengganggu proses penghitungan, misalnya data yang memiliki nilai 0.0 yang mana dapat mengganggu penghitungan numerik terhadap kolom datanya. Sehingga kita bisa melakukan preprocessing dengan cara mengganti nilai 0.0 dengan NaN, kemudian kita juga dapat mengecilkan rentangnya dengan normalisasi baik dengan MinMax ataupun ZScore. Tujuannya tentu tidak lebih dari pembersihan data yang dapat berupa penghapusan data noisy, data yang tidak konsisten ataupun membersihakn pencilan data

#### 2. 

# 5. Berikan contoh aplikasi hasil dari pembelajaran mesin!

Berikut contoh dari aplikasi pembelajaran mesin
#### 1. Bot dalam permainan video game

Dalam permainan video game yang bersifat single player, biasanya dalam dunia game tersebut juga terdapat player lain namun tidak dikendalikan oleh manusia yang mana kita kenal dengan BOT, Non Playable Character (NPC) ataupun komputer (Pada kebanyakan game terutama permainan yang bergenre Fighting atau pertarungan kerap kali memakai nama "Computer" atau CPU pada bot). Apabila diperhatikan gerakan yang dilakukan oleh Bot permainan video game ini sendiri bergerak mengikuti pola yang telah dibuat oleh Developer dari game tersebut. Kebanyakan Bot yang ada pada permainan video game menurut saya NPC-nya sangat kompleks dimana setiap role atau peran dari NPC dalam game sering kali memiliki pola yang berbeda dengan yang lainnnya, bahkan juga merespon terhadap tindakan yang dilakukan oleh player. Contoh dari Bot dalam permainan video game adalah OpenAI pada permainan DOTA 2 dimana bisa mengalahkan pemain papan atas.

#### 2. Speech Recognition

Pada beberapa perangkat android ataupun IOS, kita dapat menemukan sebuah opsi dimana kita tidak perlu mengetikan apa yang akan kita cari pada internet. Dengan begitu kita dapat memanfaatkan suara kita sebagai inputan yang nantinya diterima oleh ponsel pintar tersebut untuk dijadikan input. Pada perangkat Android terdapat aplikasi untuk mengolah suara ini yakni Google Speech API, dengan aplikasi ini kita cukup mengatakan "Hey Google" ataupun dengan menekan lama tombol home untuk menginisiasi aplikasi ini kemudian muncul dan mempelajari apa yang kita ucapkan dan membuatnya menjadi sebuah teks (kebenaran dari apa yang dicari sesuai dengan tingkat kejelasan kita dalam mengatakannya). Teks tersebut akan dicari oleh aplikasi ini dan memunculkan hasil yang relevan dengan pencarian kita.

#### 3. Algoritma rekomendasi pada website internet

Pada kebanyakan halaman yang ada pada internet, seperti youtube maupun aplikasi e-commerce. Kita dapat melihat akan muncul suatu pilihan yang mirip dengan apa yang kita cari sebelumnya, seperti pada youtube misalnya, ketika saya sering mencari informasi mengenai COVID maka pada suatu saat kedepannya youtube itu akan menampilkan video yang berisikan informasi mengenai COVID pada halaman home sedangkan saya tidak melakukan pencarian apapun. Begitu juga dengan e-commerce, ketika kita sering mencari suatu produk maka suatu saat kedepannya halaman home pada e-commerce tersebut juga akan menampilkan produk-produk yang mirip dengan produk yang kita cari sebelumnya. 

# 6. Dataset
Bukalah dataset berikut : https://archive.ics.uci.edu/ml/datasets/Dry+Bean+Dataset
Ambillah 2 sampel pertama dari masing-masing kelas sebagai data uji, sedangkan sisanya sebagai data
latih. Lakukanlah pembelajaran mesin menggunakan Naive Bayes, k-nearest Neighbour (k=3, k=7, k=11,
distance=euclidean distance), Decision trees (pohon keputusan) menggunakan Gain ratio. Kemudian
bandingkanlah akurasi diantara ketiga algoritma tersebut menggunakan pemrosesan awal PCA (principal
component analysis) maupun tidak. Berikanlah analisa dan kesimpulannya

## Proses memperoleh dataset

File ipynb ini dapat di reset factory terlebih dahulu untuk mencegah adanya error atau galat ketika proses penghitungan.

Berikut kode program untuk memperoleh dataset dari repository.

In [1]:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/00602/DryBeanDataset.zip

--2021-04-13 10:50:39--  https://archive.ics.uci.edu/ml/machine-learning-databases/00602/DryBeanDataset.zip
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 4738776 (4.5M) [application/x-httpd-php]
Saving to: ‘DryBeanDataset.zip’

DryBeanDataset.zip  100%[===================>]   4.52M  4.31MB/s    in 1.0s    

2021-04-13 10:50:40 (4.31 MB/s) - ‘DryBeanDataset.zip’ saved [4738776/4738776]



In [2]:
# Kode berikut berfungsi untuk melakukan import terhadap library yang akan digunakan
# Beberapa library mungkin akan ditambahkan nantinya
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import scipy as sc
import csv 
import math

In [3]:
# Untuk melihat path pada pointer sekarang
!pwd 
# Untuk mengecek isi dari direktori 
!ls
# Karena file yang diperoleh adalah file ZIP, maka kita ekstrak terlebih dahulu
!unzip DryBeanDataset.zip
# 3 baris kode berikut bertujuan untuk memindahkan isi dari direktori /DryBeanDataset ke level yang lebih atas
# atau ke direktori /content
!mv /content/DryBeanDataset/Dry_Bean_Dataset.arff /content/Dry_Bean_Dataset.arff
!mv /content/DryBeanDataset/Dry_Bean_Dataset.txt /content/Dry_Bean_Dataset.txt
!mv /content/DryBeanDataset/Dry_Bean_Dataset.xlsx /content/Dry_Bean_Dataset.xlsx

/content
DryBeanDataset.zip  sample_data
Archive:  DryBeanDataset.zip
   creating: DryBeanDataset/
  inflating: DryBeanDataset/Dry_Bean_Dataset.arff  
  inflating: DryBeanDataset/Dry_Bean_Dataset.txt  
  inflating: DryBeanDataset/Dry_Bean_Dataset.xlsx  


In [4]:
#Cek apakah file tadi sudah dipindahkan ke direktori /content atau belum
!ls

DryBeanDataset	       Dry_Bean_Dataset.txt   DryBeanDataset.zip
Dry_Bean_Dataset.arff  Dry_Bean_Dataset.xlsx  sample_data


Dari percobaan diatas dapat terlihat data yang ada dari repository tersebut hanya format arff, txt dan xlsx. Apabila semua data dicek dengan perintah CAT maka dapat dilihat juga dataset berada pada file dengan format xlsx yang mana merupakan file excel yang tidak dapat diproses langsung dengan python. Sehingga kita perlu mentransformasikannya menjadi file dengan format yang didukung oleh python, misalnya csv. Berikut adalah kode program untuk mentransformasikan file xlsx menjadi csv.

In [5]:
read_file = pd.read_excel (r'Dry_Bean_Dataset.xlsx')
read_file.to_csv (r'Dry_Bean_Dataset.csv', index = None, header=True)

In [6]:
# Kode berikut untuk melihat isi
!cat Dry_Bean_Dataset.csv

Streaming output truncated to the last 5000 lines.
179387,1617.553,621.6396128733878,370.2418889962433,1.679009402633248,0.8032891200682245,181149,477.9148692093672,0.8222613367069577,0.9902732005144936,0.8615565569965123,0.7687973213294989,0.0034653548633590383,0.000746749469950282,0.5910493212834129,0.9923773678548499,BOMBAY
179474,1655.841,611.4539473114847,377.6808177349781,1.6189700895546864,0.7864321791154899,183984,478.03074592728717,0.7880653376657591,0.9754869988694669,0.8225724171149987,0.7817935398555378,0.003406922157590987,0.0007850734370367198,0.6112011389598524,0.9895163626777587,BOMBAY
179496,1606.14,594.6168787506103,386.0202109982423,1.540377580782468,0.7606250216093461,181511,478.06004363655273,0.78140263810892,0.9888987444287123,0.8743752233633592,0.8039799419098849,0.0033127026716506794,0.000853774297166423,0.646383746993422,0.9956750324508343,BOMBAY
179725,1592.405,591.5388324167595,388.9329689046206,1.5209274597696156,0.7534601696351877,181562,478.3648996085803,0

IOPub data rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_data_rate_limit`.

Current values:
NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)
NotebookApp.rate_limit_window=3.0 (secs)



In [7]:
# Berikut adalah kode program untuk mengambil data dan memasukannya ke dalam dataframe
df_bean = pd.read_csv('Dry_Bean_Dataset.csv')

In [8]:
df_bean

Area  Perimeter  MajorAxisLength  ...  ShapeFactor3  ShapeFactor4     Class
0      28395    610.291       208.178117  ...      0.834222      0.998724     SEKER
1      28734    638.018       200.524796  ...      0.909851      0.998430     SEKER
2      29380    624.110       212.826130  ...      0.825871      0.999066     SEKER
3      30008    645.884       210.557999  ...      0.861794      0.994199     SEKER
4      30140    620.134       201.847882  ...      0.941900      0.999166     SEKER
...      ...        ...              ...  ...           ...           ...       ...
13606  42097    759.696       288.721612  ...      0.642988      0.998385  DERMASON
13607  42101    757.499       281.576392  ...      0.676099      0.998219  DERMASON
13608  42139    759.321       281.539928  ...      0.676884      0.996767  DERMASON
13609  42147    763.779       283.382636  ...      0.668237      0.995222  DERMASON
13610  42159    772.237       295.142741  ...      0.616221      0.998180  DERMASON

[13611 rows x 17 columns]

Setelah penjalanan beberapa blok kode program diatas, diperoleh 13611 data dengan setiap data memiliki 17 macam atribut. Sehingga dengan demikian dapat dilihat dari halaman https://archive.ics.uci.edu/ml/datasets/Dry+Bean+Dataset sendiri juga dinyatakan ada 13611 data dengan atribut sebanyak 17 macam. Sehingga data yang ada sudah benar dan dapat kita lanjutkan kepada tahap berikutnya. Sebelum dilanjutkan kita dapat mengecek apakah ada data yang fiturnya bernilai 0.0 sehingga jika perlu nanti dilakukan imputasi agar tidak mempengaruhi penghitungan data.

In [9]:
df_bean.columns

Index(['Area', 'Perimeter', 'MajorAxisLength', 'MinorAxisLength',
       'AspectRation', 'Eccentricity', 'ConvexArea', 'EquivDiameter', 'Extent',
       'Solidity', 'roundness', 'Compactness', 'ShapeFactor1', 'ShapeFactor2',
       'ShapeFactor3', 'ShapeFactor4', 'Class'],
      dtype='object')

In [10]:
df_bean.isnull().values.any()

False

Karena data diats tidak terdapat data yang nilainya 0.0 seperti hasil penjalanan kode program diatas, maka kita dapat langsung mengambil data uji dan data latihnya.

In [11]:
df_bean['Class'].unique()

array(['SEKER', 'BARBUNYA', 'BOMBAY', 'CALI', 'HOROZ', 'SIRA', 'DERMASON'],
      dtype=object)

Dari penjalanan kode program diatas, terlihat bahwa ada 7 kelas dari Dry Bean ini yakni Seker, Barbunya, Bombay, Cali, Horoz, Sira dan Dermason. Maka kita harus menghitung prior dari setiap kelas ini terlebih dahulu untuk dapat melakukan klasifikasi Naive Bayes. Karena sebelumnya sudah pernah dilakukan normalisasi maka saya akan menggunakan dataset yang sudah dinormalisasi tersebut untuk memudahkan penghitungan Naive Bayes.

In [12]:
print('Index dari kacang yang kelasnya SEKER adalah', df_bean[df_bean['Class']=='SEKER'].index.values)
print('Index dari kacang yang kelasnya BARBUNYA adalah',df_bean[df_bean['Class']=='BARBUNYA'].index.values)
print('Index dari kacang yang kelasnya BOMBAY adalah',df_bean[df_bean['Class']=='BOMBAY'].index.values)
print('Index dari kacang yang kelasnya CALI adalah',df_bean[df_bean['Class']=='CALI'].index.values)
print('Index dari kacang yang kelasnya HOROZ adalah',df_bean[df_bean['Class']=='HOROZ'].index.values)
print('Index dari kacang yang kelasnya SIRA adalah',df_bean[df_bean['Class']=='SIRA'].index.values)
print('Index dari kacang yang kelasnya DERMASON adalah',df_bean[df_bean['Class']=='DERMASON'].index.values)


Index dari kacang yang kelasnya SEKER adalah [   0    1    2 ... 2024 2025 2026]
Index dari kacang yang kelasnya BARBUNYA adalah [2027 2028 2029 ... 3346 3347 3348]
Index dari kacang yang kelasnya BOMBAY adalah [3349 3350 3351 3352 3353 3354 3355 3356 3357 3358 3359 3360 3361 3362
 3363 3364 3365 3366 3367 3368 3369 3370 3371 3372 3373 3374 3375 3376
 3377 3378 3379 3380 3381 3382 3383 3384 3385 3386 3387 3388 3389 3390
 3391 3392 3393 3394 3395 3396 3397 3398 3399 3400 3401 3402 3403 3404
 3405 3406 3407 3408 3409 3410 3411 3412 3413 3414 3415 3416 3417 3418
 3419 3420 3421 3422 3423 3424 3425 3426 3427 3428 3429 3430 3431 3432
 3433 3434 3435 3436 3437 3438 3439 3440 3441 3442 3443 3444 3445 3446
 3447 3448 3449 3450 3451 3452 3453 3454 3455 3456 3457 3458 3459 3460
 3461 3462 3463 3464 3465 3466 3467 3468 3469 3470 3471 3472 3473 3474
 3475 3476 3477 3478 3479 3480 3481 3482 3483 3484 3485 3486 3487 3488
 3489 3490 3491 3492 3493 3494 3495 3496 3497 3498 3499 3500 3501 3502
 3503 35

### 6.1 K-Nearest Neighboor dan Normalisasi Data

Sebelum kita menghitung K-Nearest Neighboor ataupun metode algoritma yang lainnya. Kita dapat melakukan pemrosesan data awal dulu agar nantinya kita dapat lebih mudah dalam menghitung datanya, dalam hal ini yang dapat dilakukan adalah melakukan normalisasi yang mana dapat memudahkan proses penghitungan baik untuk algoritma KNN, Naive Bayes ataupun Decision Tree. Untuk itu blok kode program berikut berfungsi untuk menampilkan rentang fitur dari dataset Dry Bean kecuali fitur terakhirnya, karena fitur terakhir merupakan fitur yang bersifat kualitatif sehingga tidak dapat dihitung. Blok kode program berikut berfungsi untuk menghitung rentang fitur dari semua data.

In [13]:
def rentang(input):
  print('Berikut adalah rentang dari masing-masing fitur :')
  daftar_fitur=input.columns[:-1] # Karena bagian terakhir adalah kelas yang mana berupa string sehingga tidak bisa dihitung dengan numerik
  for fitur in daftar_fitur:
    max=input[fitur].max()
    min=input[fitur].min()
    print("Rentang fitur", fitur, "Adalah", max-min)

Sedangkan blok kode berikut adalah kode program untuk melakukan tahapan normalisasi data dengan memanfaatkan library bawaan yang dimiliki oleh Sklearn

In [14]:
kolomfitur = ['Area', 'Perimeter', 'MajorAxisLength', 'MinorAxisLength',
       'AspectRation', 'Eccentricity', 'ConvexArea', 'EquivDiameter', 'Extent',
       'Solidity', 'roundness', 'Compactness', 'ShapeFactor1', 'ShapeFactor2',
       'ShapeFactor3', 'ShapeFactor4'] #kolom class tidak termasuk karena merupakan data kualitatif
from sklearn.preprocessing import MinMaxScaler
Norm = MinMaxScaler()
data_normalisasi = df_bean
data_normalisasi[kolomfitur] = Norm.fit_transform(data_normalisasi[kolomfitur])

In [121]:
data_uji = pd.DataFrame(columns=data_normalisasi.columns)
#Kode berikut untuk mengambil index 0 dan 1 yang mana pada dataframe memiliki kelas SEKER
data_uji = data_uji.append(data_normalisasi.iloc[0])
data_uji = data_uji.append(data_normalisasi.iloc[1])
#Kode berikut untuk mengambil index 0 dan 1 yang mana pada dataframe memiliki kelas BARBUNYA
data_uji = data_uji.append(data_normalisasi.iloc[2027])
data_uji = data_uji.append(data_normalisasi.iloc[2028])
#Kode berikut untuk mengambil index 0 dan 1 yang mana pada dataframe memiliki kelas BOMBAY
data_uji = data_uji.append(data_normalisasi.iloc[3349])
data_uji = data_uji.append(data_normalisasi.iloc[3350])
#Kode berikut untuk mengambil index 0 dan 1 yang mana pada dataframe memiliki kelas CALI
data_uji = data_uji.append(data_normalisasi.iloc[3871])
data_uji = data_uji.append(data_normalisasi.iloc[3872])
#Kode berikut untuk mengambil index 0 dan 1 yang mana pada dataframe memiliki kelas HOROZ
data_uji = data_uji.append(data_normalisasi.iloc[5501])
data_uji = data_uji.append(data_normalisasi.iloc[5502])
#Kode berikut untuk mengambil index 0 dan 1 yang mana pada dataframe memiliki kelas SIRA
data_uji = data_uji.append(data_normalisasi.iloc[7429])
data_uji = data_uji.append(data_normalisasi.iloc[7430])
#Kode berikut untuk mengambil index 0 dan 1 yang mana pada dataframe memiliki kelas DERMASON
data_uji = data_uji.append(data_normalisasi.iloc[10065])
data_uji = data_uji.append(data_normalisasi.iloc[10066])

In [122]:
data_latih = data_normalisasi
#14 baris kode program dibawah berfunsi untuk menghapus data yang telah diambil sebagai data uji
data_latih = data_latih.drop(0)
data_latih = data_latih.drop(1)
data_latih = data_latih.drop(2027)
data_latih = data_latih.drop(2028)
data_latih = data_latih.drop(3349)
data_latih = data_latih.drop(3350)
data_latih = data_latih.drop(3871)
data_latih = data_latih.drop(3872)
data_latih = data_latih.drop(5501)
data_latih = data_latih.drop(5502)
data_latih = data_latih.drop(7429)
data_latih = data_latih.drop(7430)
data_latih = data_latih.drop(10065)
data_latih = data_latih.drop(10066)

In [123]:
data_uji

Area  Perimeter  ...  ShapeFactor4     Class
0      0.034053   0.058574  ...      0.980620     SEKER
1      0.035500   0.077557  ...      0.974979     SEKER
2027   0.089955   0.199341  ...      0.956965  BARBUNYA
2028   0.089959   0.160763  ...      0.961797  BARBUNYA
3349   0.399597   0.516639  ...      0.877214    BOMBAY
3350   0.412535   0.507444  ...      0.930827    BOMBAY
3871   0.107107   0.183948  ...      0.983035      CALI
3872   0.107799   0.181451  ...      0.973565      CALI
5501   0.053741   0.127178  ...      0.859029     HOROZ
5502   0.054839   0.133222  ...      0.704410     HOROZ
7429   0.047392   0.103999  ...      0.947823      SIRA
7430   0.048383   0.100763  ...      0.673435      SIRA
10065  0.000000   0.000134  ...      0.967580  DERMASON
10066  0.000188   0.002514  ...      0.986868  DERMASON

[14 rows x 17 columns]

In [124]:
data_latih

Area  Perimeter  ...  ShapeFactor4     Class
2      0.038259   0.068035  ...      0.987196     SEKER
3      0.040940   0.082942  ...      0.893675     SEKER
4      0.041504   0.065313  ...      0.989116     SEKER
5      0.042097   0.075441  ...      0.990455     SEKER
6      0.042943   0.099475  ...      0.986862     SEKER
...         ...        ...  ...           ...       ...
13606  0.092559   0.160862  ...      0.974113  DERMASON
13607  0.092576   0.159358  ...      0.970912  DERMASON
13608  0.092739   0.160605  ...      0.943025  DERMASON
13609  0.092773   0.163657  ...      0.913342  DERMASON
13610  0.092824   0.169448  ...      0.970162  DERMASON

[13597 rows x 17 columns]

In [18]:
data_latih

Area  Perimeter  ...  ShapeFactor4     Class
2      0.038259   0.068035  ...      0.987196     SEKER
3      0.040940   0.082942  ...      0.893675     SEKER
4      0.041504   0.065313  ...      0.989116     SEKER
5      0.042097   0.075441  ...      0.990455     SEKER
6      0.042943   0.099475  ...      0.986862     SEKER
...         ...        ...  ...           ...       ...
13606  0.092559   0.160862  ...      0.974113  DERMASON
13607  0.092576   0.159358  ...      0.970912  DERMASON
13608  0.092739   0.160605  ...      0.943025  DERMASON
13609  0.092773   0.163657  ...      0.913342  DERMASON
13610  0.092824   0.169448  ...      0.970162  DERMASON

[13597 rows x 17 columns]

In [19]:
print("Banyak data aslinya adalah",df_bean.shape[0])
print("Banyak data latih adalah",data_latih.shape[0])
print("Banyak data uji adalah",data_uji.shape[0])

Banyak data aslinya adalah 13611
Banyak data latih adalah 13597
Banyak data uji adalah 14


In [20]:
rentang(df_bean)

Berikut adalah rentang dari masing-masing fitur :
Rentang fitur Area Adalah 1.0
Rentang fitur Perimeter Adalah 0.9999999999999998
Rentang fitur MajorAxisLength Adalah 0.9999999999999999
Rentang fitur MinorAxisLength Adalah 1.0
Rentang fitur AspectRation Adalah 1.0
Rentang fitur Eccentricity Adalah 1.0
Rentang fitur ConvexArea Adalah 1.0
Rentang fitur EquivDiameter Adalah 1.0
Rentang fitur Extent Adalah 1.0000000000000002
Rentang fitur Solidity Adalah 1.0
Rentang fitur roundness Adalah 1.0
Rentang fitur Compactness Adalah 1.0
Rentang fitur ShapeFactor1 Adalah 1.0
Rentang fitur ShapeFactor2 Adalah 1.0
Rentang fitur ShapeFactor3 Adalah 0.9999999999999999
Rentang fitur ShapeFactor4 Adalah 1.0


Dapat dilihat rentang fiturnya lumayan besar dan hal itu wajar, karena fitur yang memiliki ukuran besar tersebut adalah hasil operasi matematis seperti Area/Luas yang mana diperoleh dari hasil perkalian sehingga tentu nilainya akan besar. Oleh karenanya saya melakukan normalisasi dengan menggunakan metode MinMax dari library sklearn yakni MinManScaler. Kode programnya sendiri seperti blok kode program berikut.

In [21]:
# kode berikut untuk melihat ringkasan isi dari dataset yang sudah dinormalisasi
data_normalisasi

Area  Perimeter  ...  ShapeFactor4     Class
0      0.034053   0.058574  ...      0.980620     SEKER
1      0.035500   0.077557  ...      0.974979     SEKER
2      0.038259   0.068035  ...      0.987196     SEKER
3      0.040940   0.082942  ...      0.893675     SEKER
4      0.041504   0.065313  ...      0.989116     SEKER
...         ...        ...  ...           ...       ...
13606  0.092559   0.160862  ...      0.974113  DERMASON
13607  0.092576   0.159358  ...      0.970912  DERMASON
13608  0.092739   0.160605  ...      0.943025  DERMASON
13609  0.092773   0.163657  ...      0.913342  DERMASON
13610  0.092824   0.169448  ...      0.970162  DERMASON

[13611 rows x 17 columns]

In [22]:
# kode program berikut untuk melihat rentang dari data yang sudah dinormalisasi
rentang(data_normalisasi)

Berikut adalah rentang dari masing-masing fitur :
Rentang fitur Area Adalah 1.0
Rentang fitur Perimeter Adalah 0.9999999999999998
Rentang fitur MajorAxisLength Adalah 0.9999999999999999
Rentang fitur MinorAxisLength Adalah 1.0
Rentang fitur AspectRation Adalah 1.0
Rentang fitur Eccentricity Adalah 1.0
Rentang fitur ConvexArea Adalah 1.0
Rentang fitur EquivDiameter Adalah 1.0
Rentang fitur Extent Adalah 1.0000000000000002
Rentang fitur Solidity Adalah 1.0
Rentang fitur roundness Adalah 1.0
Rentang fitur Compactness Adalah 1.0
Rentang fitur ShapeFactor1 Adalah 1.0
Rentang fitur ShapeFactor2 Adalah 1.0
Rentang fitur ShapeFactor3 Adalah 0.9999999999999999
Rentang fitur ShapeFactor4 Adalah 1.0


Dapat dilihat setelah melakukan normalisasi, nilai dari semua fitur sudah berada pada rentang 0 hingga 1. Namun pada fitur Extent mungkin masih terdapat pencilan data darena datanya terlihat lebih 0.0000000000000002. Namun pada dasarnya hal tersebut tidak masalah karena pada kenyataannya dengan metode normalisasi MinMax memang hanya memperkecil rentang namun tidak mengatasi pencilan data.  Dibandingkan dengan data aslinya yang rentangnya lumayan besar yang mana hal itu wajar karena seperti pada nama parameternya adalah Area atau Luas dan Perimter atau keliling, sehingga tidak mungkni berada dalam rentang 0 hingga 1. Dengan normalisasi nantinya rentangnya akan berubah menjadui antara 0 dan 1 namun perbandingan luasnya tetap sama setiap datanya sehingga tidak merubah representasi dari luas ataupun atribut lain yang memiliki nilai yang sanag besar pada data aslinya.

Karena permintaan pada soal adalah dengan menggunakan Euclidean, maka berikut kodingan untuk rumus menghitung jarak euclidean beserta rumus KNN. Selain itu karena pada penghitungan KNN sendiri hanya dapat menghitung fitur yang bersifat numerikm, maka kita dapat memisahkan label yang mana pada kasus ini adalah fitur Kelas dengan data latih dan data uji dengan cara melakukan pop terhadap datanya yakni seperti blok kode program berikut.

In [23]:
label_latih = data_latih.pop('Class')
label_uji = data_uji.pop('Class')

In [24]:
def euclidean(input_1,input_2):
    return np.sqrt(np.sum(np.square(input_1-input_2)))

In [25]:
# rumus KNN
from collections import Counter
def knn(k,data_latih,label,data_uji):
    jarak = np.array([euclidean(data_latih.iloc[x],data_uji) for x in range(data_latih.shape[0])])
    min_k = jarak.argsort()[:k]
    k_kelas = label.iloc[min_k].to_numpy()
    hitung = Counter(k_kelas)
    kelas_terdekat = hitung.most_common(1)[0][0]
    return kelas_terdekat

Proses berikut akan memakan waktu lumayan lama karena menguji 14 data dengan setiap satu data uji diujikan dengan data latih yang jumlahnya 13597 buah.

In [26]:
print("KELAS DATA UJI = SEKER")
print("Berikut adalah kelas yang lebih dekat dengan data uji pertama")
kelas=knn(3,data_latih,label_latih,data_uji.iloc[0])
print("Kelas ketika k bernilai 3 adalah",kelas)
kelas=knn(7,data_latih,label_latih,data_uji.iloc[0])
print("Kelas ketika k bernilai 7 adalah",kelas)
kelas=knn(11,data_latih,label_latih,data_uji.iloc[0])
print("Kelas ketika k bernilai 11 adalah",kelas)
print()
print("Berikut adalah kelas yang lebih dekat dengan data uji kedua")
kelas=knn(3,data_latih,label_latih,data_uji.iloc[1])
print("Kelas ketika k bernilai 3 adalah",kelas)
kelas=knn(7,data_latih,label_latih,data_uji.iloc[1])
print("Kelas ketika k bernilai 7 adalah",kelas)
kelas=knn(11,data_latih,label_latih,data_uji.iloc[1])
print("Kelas ketika k bernilai 11 adalah",kelas)
print()

KELAS DATA UJI = SEKER
Berikut adalah kelas yang lebih dekat dengan data uji pertama
Kelas ketika k bernilai 3 adalah DERMASON
Kelas ketika k bernilai 7 adalah SEKER
Kelas ketika k bernilai 11 adalah SEKER

Berikut adalah kelas yang lebih dekat dengan data uji kedua
Kelas ketika k bernilai 3 adalah SEKER
Kelas ketika k bernilai 7 adalah SEKER
Kelas ketika k bernilai 11 adalah SEKER



In [27]:
print("KELAS DATA UJI = BARBUNYA")
print("Berikut adalah kelas yang lebih dekat dengan data uji ketiga")
kelas=knn(3,data_latih,label_latih,data_uji.iloc[2])
print("Kelas ketika k bernilai 3 adalah",kelas)
kelas=knn(7,data_latih,label_latih,data_uji.iloc[2])
print("Kelas ketika k bernilai 7 adalah",kelas)
kelas=knn(11,data_latih,label_latih,data_uji.iloc[2])
print("Kelas ketika k bernilai 11 adalah",kelas)
print()
print("Berikut adalah kelas yang lebih dekat dengan data uji keempat")
kelas=knn(3,data_latih,label_latih,data_uji.iloc[3])
print("Kelas ketika k bernilai 3 adalah",kelas)
kelas=knn(7,data_latih,label_latih,data_uji.iloc[3])
print("Kelas ketika k bernilai 7 adalah",kelas)
kelas=knn(11,data_latih,label_latih,data_uji.iloc[3])
print("Kelas ketika k bernilai 11 adalah",kelas)
print()

KELAS DATA UJI = BARBUNYA
Berikut adalah kelas yang lebih dekat dengan data uji pertama
Kelas ketika k bernilai 3 adalah SIRA
Kelas ketika k bernilai 7 adalah SIRA
Kelas ketika k bernilai 11 adalah SIRA

Berikut adalah kelas yang lebih dekat dengan data uji kedua
Kelas ketika k bernilai 3 adalah SIRA
Kelas ketika k bernilai 7 adalah SIRA
Kelas ketika k bernilai 11 adalah SIRA



In [28]:
print("KELAS DATA UJI = BOMBAY")
print("Berikut adalah kelas yang lebih dekat dengan data uji kelima")
kelas=knn(3,data_latih,label_latih,data_uji.iloc[4])
print("Kelas ketika k bernilai 3 adalah",kelas)
kelas=knn(7,data_latih,label_latih,data_uji.iloc[4])
print("Kelas ketika k bernilai 7 adalah",kelas)
kelas=knn(11,data_latih,label_latih,data_uji.iloc[4])
print("Kelas ketika k bernilai 11 adalah",kelas)
print()
print("Berikut adalah kelas yang lebih dekat dengan data uji kenam")
kelas=knn(3,data_latih,label_latih,data_uji.iloc[5])
print("Kelas ketika k bernilai 3 adalah",kelas)
kelas=knn(7,data_latih,label_latih,data_uji.iloc[5])
print("Kelas ketika k bernilai 7 adalah",kelas)
kelas=knn(11,data_latih,label_latih,data_uji.iloc[5])
print("Kelas ketika k bernilai 11 adalah",kelas)
print()

KELAS DATA UJI = BOMBAY
Berikut adalah kelas yang lebih dekat dengan data uji pertama
Kelas ketika k bernilai 3 adalah BOMBAY
Kelas ketika k bernilai 7 adalah BOMBAY
Kelas ketika k bernilai 11 adalah BOMBAY

Berikut adalah kelas yang lebih dekat dengan data uji kedua
Kelas ketika k bernilai 3 adalah BOMBAY
Kelas ketika k bernilai 7 adalah BOMBAY
Kelas ketika k bernilai 11 adalah BOMBAY



In [29]:
print("KELAS DATA UJI = CALI")
print("Berikut adalah kelas yang lebih dekat dengan data uji ketujuh")
kelas=knn(3,data_latih,label_latih,data_uji.iloc[6])
print("Kelas ketika k bernilai 3 adalah",kelas)
kelas=knn(7,data_latih,label_latih,data_uji.iloc[6])
print("Kelas ketika k bernilai 7 adalah",kelas)
kelas=knn(11,data_latih,label_latih,data_uji.iloc[6])
print("Kelas ketika k bernilai 11 adalah",kelas)
print()
print("Berikut adalah kelas yang lebih dekat dengan data uji kedelapan")
kelas=knn(3,data_latih,label_latih,data_uji.iloc[7])
print("Kelas ketika k bernilai 3 adalah",kelas)
kelas=knn(7,data_latih,label_latih,data_uji.iloc[7])
print("Kelas ketika k bernilai 7 adalah",kelas)
kelas=knn(11,data_latih,label_latih,data_uji.iloc[7])
print("Kelas ketika k bernilai 11 adalah",kelas)
print()

KELAS DATA UJI = CALI
Berikut adalah kelas yang lebih dekat dengan data uji pertama
Kelas ketika k bernilai 3 adalah SIRA
Kelas ketika k bernilai 7 adalah SIRA
Kelas ketika k bernilai 11 adalah SIRA

Berikut adalah kelas yang lebih dekat dengan data uji kedua
Kelas ketika k bernilai 3 adalah SEKER
Kelas ketika k bernilai 7 adalah SEKER
Kelas ketika k bernilai 11 adalah SEKER



In [30]:
print("KELAS DATA UJI = HOROZ")
print("Berikut adalah kelas yang lebih dekat dengan data uji kesembilan")
kelas=knn(3,data_latih,label_latih,data_uji.iloc[8])
print("Kelas ketika k bernilai 3 adalah",kelas)
kelas=knn(7,data_latih,label_latih,data_uji.iloc[8])
print("Kelas ketika k bernilai 7 adalah",kelas)
kelas=knn(11,data_latih,label_latih,data_uji.iloc[8])
print("Kelas ketika k bernilai 11 adalah",kelas)
print()
print("Berikut adalah kelas yang lebih dekat dengan data uji kesepuluh")
kelas=knn(3,data_latih,label_latih,data_uji.iloc[9])
print("Kelas ketika k bernilai 3 adalah",kelas)
kelas=knn(7,data_latih,label_latih,data_uji.iloc[9])
print("Kelas ketika k bernilai 7 adalah",kelas)
kelas=knn(11,data_latih,label_latih,data_uji.iloc[9])
print("Kelas ketika k bernilai 11 adalah",kelas)
print()

KELAS DATA UJI = HOROZ
Berikut adalah kelas yang lebih dekat dengan data uji pertama
Kelas ketika k bernilai 3 adalah HOROZ
Kelas ketika k bernilai 7 adalah HOROZ
Kelas ketika k bernilai 11 adalah HOROZ

Berikut adalah kelas yang lebih dekat dengan data uji kedua
Kelas ketika k bernilai 3 adalah DERMASON
Kelas ketika k bernilai 7 adalah DERMASON
Kelas ketika k bernilai 11 adalah DERMASON



In [31]:
print("KELAS DATA UJI = SIRA")
print("Berikut adalah kelas yang lebih dekat dengan data uji kesebelas")
kelas=knn(3,data_latih,label_latih,data_uji.iloc[10])
print("Kelas ketika k bernilai 3 adalah",kelas)
kelas=knn(7,data_latih,label_latih,data_uji.iloc[10])
print("Kelas ketika k bernilai 7 adalah",kelas)
kelas=knn(11,data_latih,label_latih,data_uji.iloc[10])
print("Kelas ketika k bernilai 11 adalah",kelas)
print()
print("Berikut adalah kelas yang lebih dekat dengan data uji keduabelas")
kelas=knn(3,data_latih,label_latih,data_uji.iloc[11])
print("Kelas ketika k bernilai 3 adalah",kelas)
kelas=knn(7,data_latih,label_latih,data_uji.iloc[11])
print("Kelas ketika k bernilai 7 adalah",kelas)
kelas=knn(11,data_latih,label_latih,data_uji.iloc[11])
print("Kelas ketika k bernilai 11 adalah",kelas)
print()

KELAS DATA UJI = SIRA
Berikut adalah kelas yang lebih dekat dengan data uji pertama
Kelas ketika k bernilai 3 adalah DERMASON
Kelas ketika k bernilai 7 adalah DERMASON
Kelas ketika k bernilai 11 adalah DERMASON

Berikut adalah kelas yang lebih dekat dengan data uji kedua
Kelas ketika k bernilai 3 adalah DERMASON
Kelas ketika k bernilai 7 adalah DERMASON
Kelas ketika k bernilai 11 adalah DERMASON



In [32]:
print("KELAS DATA UJI = DERMASON")
print("Berikut adalah kelas yang lebih dekat dengan data uji ketigabelas")
kelas=knn(3,data_latih,label_latih,data_uji.iloc[12])
print("Kelas ketika k bernilai 3 adalah",kelas)
kelas=knn(7,data_latih,label_latih,data_uji.iloc[12])
print("Kelas ketika k bernilai 7 adalah",kelas)
kelas=knn(11,data_latih,label_latih,data_uji.iloc[12])
print("Kelas ketika k bernilai 11 adalah",kelas)
print()
print("Berikut adalah kelas yang lebih dekat dengan data uji keempatbelas")
kelas=knn(3,data_latih,label_latih,data_uji.iloc[13])
print("Kelas ketika k bernilai 3 adalah",kelas)
kelas=knn(7,data_latih,label_latih,data_uji.iloc[13])
print("Kelas ketika k bernilai 7 adalah",kelas)
kelas=knn(11,data_latih,label_latih,data_uji.iloc[13])
print("Kelas ketika k bernilai 11 adalah",kelas)
print()

KELAS DATA UJI = DERMASON
Berikut adalah kelas yang lebih dekat dengan data uji pertama
Kelas ketika k bernilai 3 adalah DERMASON
Kelas ketika k bernilai 7 adalah DERMASON
Kelas ketika k bernilai 11 adalah DERMASON

Berikut adalah kelas yang lebih dekat dengan data uji kedua
Kelas ketika k bernilai 3 adalah DERMASON
Kelas ketika k bernilai 7 adalah DERMASON
Kelas ketika k bernilai 11 adalah DERMASON



### 6.2 Klasifikasi Naive Bayes

Naive bayes adalah sebuah algoritma yang digunakan untuk mengecek kebenaran dari suatu klasifikasi objek terhadap kelas yang sudah ditentukan. Sehingga dengan menggunakan metode Naive Bayes ini berfungsi untuk memastikan kebenarannya, apabila bernilai lebih kecil atau bahkan dibawah nol tentu tidak dekat dengan kelas tersebut. Karena dengan memprint datanya saja kita tidak dapat melihat kelas yang ada, maka kita dapat melakukan proses berikut untuk melihat kelas apa saja yang ada pada dataset Dry Bean ini.

Karena sebelumnya dilakukan pop terhadap suatu atribut yakni atribut "Kelas" pada data uji dan data latih, sedangkan pada klasifikasi Naive Bayes sendiri membutuhkan atribut kelas tersebut untuk mengelompokannya. Maka dengan begitu saya menghapus variabel data_latih dan data_uji dan kemudian membuat data_uji dan data_latih baru dengan berdasarkan kepada data yang sudah dinormalisasi sebelumnya pada tahap prepocessing di K-Nearest Neighboor.

In [125]:
del data_latih
del data_uji

In [126]:
data_uji = pd.DataFrame(columns=data_normalisasi.columns)
#Kode berikut untuk mengambil index 0 dan 1 yang mana pada dataframe memiliki kelas SEKER
data_uji = data_uji.append(data_normalisasi.iloc[0])
data_uji = data_uji.append(data_normalisasi.iloc[1])
#Kode berikut untuk mengambil index 0 dan 1 yang mana pada dataframe memiliki kelas BARBUNYA
data_uji = data_uji.append(data_normalisasi.iloc[2027])
data_uji = data_uji.append(data_normalisasi.iloc[2028])
#Kode berikut untuk mengambil index 0 dan 1 yang mana pada dataframe memiliki kelas BOMBAY
data_uji = data_uji.append(data_normalisasi.iloc[3349])
data_uji = data_uji.append(data_normalisasi.iloc[3350])
#Kode berikut untuk mengambil index 0 dan 1 yang mana pada dataframe memiliki kelas CALI
data_uji = data_uji.append(data_normalisasi.iloc[3871])
data_uji = data_uji.append(data_normalisasi.iloc[3872])
#Kode berikut untuk mengambil index 0 dan 1 yang mana pada dataframe memiliki kelas HOROZ
data_uji = data_uji.append(data_normalisasi.iloc[5501])
data_uji = data_uji.append(data_normalisasi.iloc[5502])
#Kode berikut untuk mengambil index 0 dan 1 yang mana pada dataframe memiliki kelas SIRA
data_uji = data_uji.append(data_normalisasi.iloc[7429])
data_uji = data_uji.append(data_normalisasi.iloc[7430])
#Kode berikut untuk mengambil index 0 dan 1 yang mana pada dataframe memiliki kelas DERMASON
data_uji = data_uji.append(data_normalisasi.iloc[10065])
data_uji = data_uji.append(data_normalisasi.iloc[10066])

In [127]:
data_latih = data_normalisasi
#14 baris kode program dibawah berfunsi untuk menghapus data yang telah diambil sebagai data uji
data_latih = data_latih.drop(0)
data_latih = data_latih.drop(1)
data_latih = data_latih.drop(2027)
data_latih = data_latih.drop(2028)
data_latih = data_latih.drop(3349)
data_latih = data_latih.drop(3350)
data_latih = data_latih.drop(3871)
data_latih = data_latih.drop(3872)
data_latih = data_latih.drop(5501)
data_latih = data_latih.drop(5502)
data_latih = data_latih.drop(7429)
data_latih = data_latih.drop(7430)
data_latih = data_latih.drop(10065)
data_latih = data_latih.drop(10066)

In [131]:
data_latih

Area  Perimeter  ...  ShapeFactor4     Class
2      0.038259   0.068035  ...      0.987196     SEKER
3      0.040940   0.082942  ...      0.893675     SEKER
4      0.041504   0.065313  ...      0.989116     SEKER
5      0.042097   0.075441  ...      0.990455     SEKER
6      0.042943   0.099475  ...      0.986862     SEKER
...         ...        ...  ...           ...       ...
13606  0.092559   0.160862  ...      0.974113  DERMASON
13607  0.092576   0.159358  ...      0.970912  DERMASON
13608  0.092739   0.160605  ...      0.943025  DERMASON
13609  0.092773   0.163657  ...      0.913342  DERMASON
13610  0.092824   0.169448  ...      0.970162  DERMASON

[13597 rows x 17 columns]

In [128]:
data_latih.columns[-1]

'Class'

In [129]:
def hitung_prior(kelas):
  i=0
  prior = Counter(kelas)
  n_kelas = prior.keys()
  n_data = kelas.shape[0]
  for i in n_kelas:
    prior[i]=prior[i]/n_data
  return prior

In [130]:
prior = hitung_prior(data_latih['Class'])
prior

Counter({'BARBUNYA': 0.09708023828785761,
         'BOMBAY': 0.03824373023461058,
         'CALI': 0.11973229388835772,
         'DERMASON': 0.26064573067588437,
         'HOROZ': 0.14164889313819226,
         'SEKER': 0.14892991100978156,
         'SIRA': 0.19371920276531587})

In [132]:
def rata2_std_kelas(df_input):
  kolom = df_input.columns[:-1]
  kelas = df_input.columns[-1]
  daftar_kelas = set(df_input[kelas])
  rata2 = {}
  std = {}
  for k in kolom:
    for x in daftar_kelas:
      rata2[(x,k)] = df_input.loc[df_input[kelas]==x][k].mean()
      std[(x,k)] = df_input.loc[df_input[kelas]==x][k].std()
  return(rata2,std)

In [133]:
def likelihood_gaussian(df_input,rata2,std):
  rst = 1/np.sqrt(2*math.pi*(np.square(std)))
  rst = rst*np.exp((-1*(np.square(df_input-rata2)))/(2*np.square(std)))
  return rst

In [135]:
def train_bayes_gaussian(data_latih):
  kelas = data_latih.columns[-1]
  prior = hitung_prior(data_latih[kelas])
  (rata2,std) = rata2_std_kelas(data_latih)
  daftar_kelas = set(data_latih[kelas])
  daftar_kolom = data_latih.columns[:-1]
  model = {}
  model['prior'] = prior
  model['rata2'] = rata2
  model['std'] = std
  model['daftar_kelas'] = daftar_kelas
  model['daftar_kolom'] = daftar_kolom
  return model

model = train_bayes_gaussian(data_latih)

In [155]:
def testing_bayes_gaussian(model,data_uji):
  prior = model['prior']
  rata2 = model['rata2']
  std = model['std']
  daftar_kelas = model['daftar_kelas']
  kolom = model['daftar_kolom']
  posterior = dict.fromkeys(daftar_kelas,1)
  for x in daftar_kelas:
    for k in kolom:
      posterior[x] = posterior[x]*likelihood_gaussian(data_uji[k],rata2[(x,k)],std[(x,k)])
    posterior[x] = posterior[x]*prior[x]
    print('Nilai dari posterior terhadap kelas',x,'adalah :', posterior[x])
  kelas_uji = max(posterior,key=posterior.get)
  return kelas_uji

In [180]:
def uji_posterior(index):
  print('Jenis Dry Bean yang diuji adalah     :',label_uji.iloc[index])
  print("Index ke   :",index)
  print('Atribut    : ')
  print(data_uji.iloc[index])
  print()
  prediksi = testing_bayes_gaussian(model,data_uji.iloc[index])
  print()
  print('Maka       : ')
  print('Jenis Dry Bean cocok berdasarkan metode Naive Bayes adalah :',prediksi)

In [181]:
uji_posterior(0)

Jenis Dry Bean yang diuji adalah     : SEKER
Index ke : 0
Atribut : 
Area               0.0340527
Perimeter          0.0585739
MajorAxisLength    0.0442621
MinorAxisLength     0.152142
AspectRation        0.122612
Eccentricity        0.477797
ConvexArea          0.033107
EquivDiameter      0.0708041
Extent              0.671024
Solidity            0.922824
roundness           0.934823
Compactness         0.786733
ShapeFactor1        0.593432
ShapeFactor2        0.833049
ShapeFactor3        0.750996
ShapeFactor4         0.98062
Class                  SEKER
Name: 0, dtype: object

Nilai dari posterior terhadap kelas SIRA adalah : 1.0972253840506628e-41
Nilai dari posterior terhadap kelas HOROZ adalah : 2.308273840833445e-155
Nilai dari posterior terhadap kelas SEKER adalah : 1738.8273848509698
Nilai dari posterior terhadap kelas DERMASON adalah : 0.00020256633003274324
Nilai dari posterior terhadap kelas BOMBAY adalah : 3.65848219522922e-232
Nilai dari posterior terhadap kelas BARBUNYA a

In [179]:
uji_posterior(0)

Jenis Dry Bean yang diuji adalah     : SEKER
Atribut : 
Area               0.0340527
Perimeter          0.0585739
MajorAxisLength    0.0442621
MinorAxisLength     0.152142
AspectRation        0.122612
Eccentricity        0.477797
ConvexArea          0.033107
EquivDiameter      0.0708041
Extent              0.671024
Solidity            0.922824
roundness           0.934823
Compactness         0.786733
ShapeFactor1        0.593432
ShapeFactor2        0.833049
ShapeFactor3        0.750996
ShapeFactor4         0.98062
Class                  SEKER
Name: 0, dtype: object

Nilai dari posterior terhadap kelas SIRA adalah : 1.0972253840506628e-41
Nilai dari posterior terhadap kelas HOROZ adalah : 2.308273840833445e-155
Nilai dari posterior terhadap kelas SEKER adalah : 1738.8273848509698
Nilai dari posterior terhadap kelas DERMASON adalah : 0.00020256633003274324
Nilai dari posterior terhadap kelas BOMBAY adalah : 3.65848219522922e-232
Nilai dari posterior terhadap kelas BARBUNYA adalah : 3.515

In [178]:
uji_posterior(1)

Jenis Dry Bean yang diuji adalah     : SEKER
Atribut : 
Area               0.0355002
Perimeter          0.0775567
MajorAxisLength    0.0304788
MinorAxisLength     0.178337
AspectRation       0.0515774
Eccentricity        0.278472
ConvexArea          0.034991
EquivDiameter      0.0735769
Extent              0.735504
Solidity            0.871514
roundness           0.793138
Compactness         0.903549
ShapeFactor1        0.547447
ShapeFactor2        0.967315
ShapeFactor3        0.884987
ShapeFactor4        0.974979
Class                  SEKER
Name: 1, dtype: object

Nilai dari posterior terhadap kelas SIRA adalah : 3.123483862227876e-72
Nilai dari posterior terhadap kelas HOROZ adalah : 1.0578351023920896e-242
Nilai dari posterior terhadap kelas SEKER adalah : 0.2094689971236825
Nilai dari posterior terhadap kelas DERMASON adalah : 4.6656162649500175e-25
Nilai dari posterior terhadap kelas BOMBAY adalah : 2.76022050104375e-259
Nilai dari posterior terhadap kelas BARBUNYA adalah : 9.191

In [177]:
uji_posterior(2)

Jenis Dry Bean yang diuji adalah     : BARBUNYA
Atribut : 
Area               0.0899546
Perimeter           0.199341
MajorAxisLength     0.207913
MinorAxisLength     0.161596
AspectRation        0.472369
Eccentricity        0.847507
ConvexArea         0.0898642
EquivDiameter       0.168055
Extent              0.430587
Solidity            0.759753
roundness           0.585824
Compactness         0.369089
ShapeFactor1        0.577363
ShapeFactor2        0.318345
ShapeFactor3        0.319491
ShapeFactor4        0.956965
Class               BARBUNYA
Name: 2027, dtype: object

Nilai dari posterior terhadap kelas SIRA adalah : 6288.47580865709
Nilai dari posterior terhadap kelas HOROZ adalah : 0.04122663694486932
Nilai dari posterior terhadap kelas SEKER adalah : 3.30969661945352e-24
Nilai dari posterior terhadap kelas DERMASON adalah : 3.9196657739890374e-05
Nilai dari posterior terhadap kelas BOMBAY adalah : 1.800675729794265e-132
Nilai dari posterior terhadap kelas BARBUNYA adalah : 1.313

In [176]:
uji_posterior(3)

Jenis Dry Bean yang diuji adalah     : BARBUNYA
Atribut : 
Area               0.0899588
Perimeter           0.160763
MajorAxisLength     0.179271
MinorAxisLength     0.190925
AspectRation        0.348206
Eccentricity        0.768218
ConvexArea         0.0878937
EquivDiameter       0.168062
Extent              0.492534
Solidity            0.907392
roundness           0.826371
Compactness         0.493622
ShapeFactor1        0.527381
ShapeFactor2        0.407487
ShapeFactor3        0.440383
ShapeFactor4        0.961797
Class               BARBUNYA
Name: 2028, dtype: object

Nilai dari posterior terhadap kelas SIRA adalah : 11816406130899.371
Nilai dari posterior terhadap kelas HOROZ adalah : 4.784239834185022e-22
Nilai dari posterior terhadap kelas SEKER adalah : 19.059920664146553
Nilai dari posterior terhadap kelas DERMASON adalah : 127363852.45417565
Nilai dari posterior terhadap kelas BOMBAY adalah : 2.520328471877507e-122
Nilai dari posterior terhadap kelas BARBUNYA adalah : 3.06077

In [175]:
uji_posterior(4)

Jenis Dry Bean yang diuji adalah     : BOMBAY
Atribut : 
Area               0.399597
Perimeter          0.516639
MajorAxisLength    0.482226
MinorAxisLength    0.595925
AspectRation        0.26277
Eccentricity       0.690055
ConvexArea         0.390037
EquivDiameter      0.538425
Extent              0.62298
Solidity           0.921775
roundness          0.769681
Compactness        0.586968
ShapeFactor1       0.153934
ShapeFactor2       0.217885
ShapeFactor3       0.535331
ShapeFactor4       0.877214
Class                BOMBAY
Name: 3349, dtype: object

Nilai dari posterior terhadap kelas SIRA adalah : 2.0110092401035156e-221
Nilai dari posterior terhadap kelas HOROZ adalah : 2.9989221205146963e-104
Nilai dari posterior terhadap kelas SEKER adalah : 9.913999331168992e-221
Nilai dari posterior terhadap kelas DERMASON adalah : 5.363140892029526e-254
Nilai dari posterior terhadap kelas BOMBAY adalah : 2.409652577582757e-05
Nilai dari posterior terhadap kelas BARBUNYA adalah : 6.4202756236

In [174]:
uji_posterior(5)

Jenis Dry Bean yang diuji adalah     : BOMBAY
Atribut : 
Area               0.412535
Perimeter          0.507444
MajorAxisLength    0.436414
MinorAxisLength    0.677264
AspectRation       0.133657
Eccentricity       0.500772
ConvexArea         0.401254
EquivDiameter      0.550749
Extent             0.614399
Solidity           0.959916
roundness          0.854359
Compactness        0.766421
ShapeFactor1       0.112249
ShapeFactor2       0.306532
ShapeFactor3       0.728291
ShapeFactor4       0.930827
Class                BOMBAY
Name: 3350, dtype: object

Nilai dari posterior terhadap kelas SIRA adalah : 8.97204688461719e-269
Nilai dari posterior terhadap kelas HOROZ adalah : 5.86301579830204e-176
Nilai dari posterior terhadap kelas SEKER adalah : 2.1587808862279822e-234
Nilai dari posterior terhadap kelas DERMASON adalah : 1.391474452009053e-281
Nilai dari posterior terhadap kelas BOMBAY adalah : 2.11888573413701e-20
Nilai dari posterior terhadap kelas BARBUNYA adalah : 2.98138233977011

In [173]:
uji_posterior(6)

Jenis Dry Bean yang diuji adalah     : CALI
Atribut : 
Area               0.107107
Perimeter          0.183948
MajorAxisLength    0.201471
MinorAxisLength    0.218544
AspectRation       0.341699
Eccentricity       0.763094
ConvexArea         0.104247
EquivDiameter      0.194688
Extent             0.586929
Solidity           0.935602
roundness          0.835693
Compactness        0.502025
ShapeFactor1       0.484189
ShapeFactor2       0.386957
ShapeFactor3       0.448777
ShapeFactor4       0.983035
Class                  CALI
Name: 3871, dtype: object

Nilai dari posterior terhadap kelas SIRA adalah : 19670126742202.254
Nilai dari posterior terhadap kelas HOROZ adalah : 5.077238592078049e-21
Nilai dari posterior terhadap kelas SEKER adalah : 448.3332574889715
Nilai dari posterior terhadap kelas DERMASON adalah : 1701.9522535195
Nilai dari posterior terhadap kelas BOMBAY adalah : 3.52275540423887e-105
Nilai dari posterior terhadap kelas BARBUNYA adalah : 0.1541561628369465
Nilai dari pos

In [172]:
uji_posterior(7)

Jenis Dry Bean yang diuji adalah     : CALI
Atribut : 
Area               0.107799
Perimeter          0.181451
MajorAxisLength     0.16436
MinorAxisLength    0.264662
AspectRation        0.19379
Eccentricity       0.603687
ConvexArea         0.104668
EquivDiameter      0.195737
Extent             0.621021
Solidity           0.953165
roundness          0.858988
Compactness         0.68266
ShapeFactor1       0.422379
ShapeFactor2       0.527256
ShapeFactor3       0.636518
ShapeFactor4       0.973565
Class                  CALI
Name: 3872, dtype: object

Nilai dari posterior terhadap kelas SIRA adalah : 0.48948062241657525
Nilai dari posterior terhadap kelas HOROZ adalah : 2.8164492243959248e-70
Nilai dari posterior terhadap kelas SEKER adalah : 103342584552.83322
Nilai dari posterior terhadap kelas DERMASON adalah : 0.0032959843395991206
Nilai dari posterior terhadap kelas BOMBAY adalah : 2.118480747044462e-110
Nilai dari posterior terhadap kelas BARBUNYA adalah : 1.3691229288420796e-07


In [171]:
uji_posterior(8)

Jenis Dry Bean yang diuji adalah     : HOROZ
Atribut : 
Area               0.0537413
Perimeter           0.127178
MajorAxisLength      0.17905
MinorAxisLength    0.0802849
AspectRation        0.616661
Eccentricity        0.909608
ConvexArea         0.0522308
EquivDiameter       0.107209
Extent              0.257854
Solidity            0.932243
roundness           0.662621
Compactness         0.241541
ShapeFactor1        0.755465
ShapeFactor2         0.28759
ShapeFactor3        0.202521
ShapeFactor4        0.859029
Class                  HOROZ
Name: 5501, dtype: object

Nilai dari posterior terhadap kelas SIRA adalah : 3.5774074496103984e-14
Nilai dari posterior terhadap kelas HOROZ adalah : 0.0002964949152009175
Nilai dari posterior terhadap kelas SEKER adalah : 1.41650958130458e-54
Nilai dari posterior terhadap kelas DERMASON adalah : 8.709716683491388e-06
Nilai dari posterior terhadap kelas BOMBAY adalah : 4.910054267872902e-208
Nilai dari posterior terhadap kelas BARBUNYA adalah : 2

In [170]:
uji_posterior(9)

Jenis Dry Bean yang diuji adalah     : HOROZ
Atribut : 
Area               0.0548387
Perimeter           0.133222
MajorAxisLength     0.158013
MinorAxisLength     0.106766
AspectRation        0.488343
Eccentricity        0.855668
ConvexArea         0.0553391
EquivDiameter        0.10916
Extent              0.377722
Solidity            0.742126
roundness           0.635072
Compactness         0.339943
ShapeFactor1        0.701064
ShapeFactor2        0.355028
ShapeFactor3        0.292151
ShapeFactor4         0.70441
Class                  HOROZ
Name: 5502, dtype: object

Nilai dari posterior terhadap kelas SIRA adalah : 1.2388031839614133e-11
Nilai dari posterior terhadap kelas HOROZ adalah : 5.308373102018069e-10
Nilai dari posterior terhadap kelas SEKER adalah : 8.965360347170284e-53
Nilai dari posterior terhadap kelas DERMASON adalah : 1.5927755652563236e-09
Nilai dari posterior terhadap kelas BOMBAY adalah : 4.2528361793265546e-187
Nilai dari posterior terhadap kelas BARBUNYA adalah 

In [169]:
uji_posterior(10)

Jenis Dry Bean yang diuji adalah     : SIRA
Atribut : 
Area               0.0473919
Perimeter           0.103999
MajorAxisLength     0.128719
MinorAxisLength     0.104506
AspectRation        0.420894
Eccentricity        0.818388
ConvexArea         0.0469171
EquivDiameter      0.0957636
Extent              0.652078
Solidity             0.84482
roundness           0.749362
Compactness         0.417609
ShapeFactor1        0.692633
ShapeFactor2        0.430551
ShapeFactor3        0.365807
ShapeFactor4        0.947823
Class                   SIRA
Name: 7429, dtype: object

Nilai dari posterior terhadap kelas SIRA adalah : 0.012796386388457027
Nilai dari posterior terhadap kelas HOROZ adalah : 3.784786094927422e-22
Nilai dari posterior terhadap kelas SEKER adalah : 1.2134200705483805e-14
Nilai dari posterior terhadap kelas DERMASON adalah : 110561145067.3716
Nilai dari posterior terhadap kelas BOMBAY adalah : 2.834221081529948e-188
Nilai dari posterior terhadap kelas BARBUNYA adalah : 1.8200

In [168]:
uji_posterior(11)

Jenis Dry Bean yang diuji adalah     : SIRA
Atribut : 
Area               0.0483826
Perimeter           0.100763
MajorAxisLength     0.108961
MinorAxisLength      0.13625
AspectRation        0.301415
Eccentricity         0.72855
ConvexArea         0.0485537
EquivDiameter      0.0975668
Extent              0.447019
Solidity            0.780197
roundness           0.786627
Compactness         0.528106
ShapeFactor1        0.639896
ShapeFactor2        0.522045
ShapeFactor3        0.475026
ShapeFactor4        0.673435
Class                   SIRA
Name: 7430, dtype: object

Nilai dari posterior terhadap kelas SIRA adalah : 9.202639030482046e-09
Nilai dari posterior terhadap kelas HOROZ adalah : 6.216397509733834e-44
Nilai dari posterior terhadap kelas SEKER adalah : 1.717031756376938e-31
Nilai dari posterior terhadap kelas DERMASON adalah : 0.03478827915340197
Nilai dari posterior terhadap kelas BOMBAY adalah : 9.847127471699995e-181
Nilai dari posterior terhadap kelas BARBUNYA adalah : 1.23

In [182]:
uji_posterior(12)

Jenis Dry Bean yang diuji adalah     : DERMASON
Index ke : 12
Atribut : 
Area                         0
Perimeter          0.000134188
MajorAxisLength              0
MinorAxisLength      0.0573716
AspectRation          0.191496
Eccentricity          0.600318
ConvexArea                   0
EquivDiameter                0
Extent                0.755506
Solidity              0.901354
roundness             0.881353
Compactness           0.685416
ShapeFactor1          0.809736
ShapeFactor2          0.882091
ShapeFactor3           0.63949
ShapeFactor4           0.96758
Class                 DERMASON
Name: 10065, dtype: object

Nilai dari posterior terhadap kelas SIRA adalah : 1.9853502534160624e-64
Nilai dari posterior terhadap kelas HOROZ adalah : 1.0243045864216866e-139
Nilai dari posterior terhadap kelas SEKER adalah : 1.2172973463141587e-27
Nilai dari posterior terhadap kelas DERMASON adalah : 4.6590677326040916e-05
Nilai dari posterior terhadap kelas BOMBAY adalah : 0.0
Nilai dari poster

In [166]:
uji_posterior(13)

Jenis Dry Bean yang diuji adalah     : DERMASON
Atribut : 
Area               0.000187877
Perimeter           0.00251398
MajorAxisLength       0.013774
MinorAxisLength      0.0410317
AspectRation          0.268655
Eccentricity          0.696305
ConvexArea         0.000362771
EquivDiameter      0.000425419
Extent                0.617898
Solidity              0.873989
roundness             0.860934
Compactness           0.586742
ShapeFactor1          0.855923
ShapeFactor2          0.761502
ShapeFactor3          0.535096
ShapeFactor4          0.986868
Class                 DERMASON
Name: 10066, dtype: object

Nilai dari posterior terhadap kelas SIRA adalah : 1.7512053804927907e-55
Nilai dari posterior terhadap kelas HOROZ adalah : 4.091980970598743e-100
Nilai dari posterior terhadap kelas SEKER adalah : 2.512749942564458e-34
Nilai dari posterior terhadap kelas DERMASON adalah : 7.647711624708726
Nilai dari posterior terhadap kelas BOMBAY adalah : 1.13383e-319
Nilai dari posterior terhadap

In [186]:
prediksi_total =[]
for indeks in range(data_uji.shape[0]):
  prediksi_total.append(testing_naive_bayes_gaussian(model,data_uji.iloc[indeks]))
print("Total prediksi benar: ",sum(prediksi_total==label_uji))

Total prediksi benar:  7


### 6.3 Decision Tree dengan Gain Ratio

Sama seperti dua metode klasifikasi sebelumnya, pada metode Decision Tree ini dengan menggunakan Gain Ratio saya menggunakan data yang sudah dinormalisasi sebelumnya. Karena tidak terjadi proses penghapusan satu atribut maka data uji dan data latih yang sebelumnya digunakan pada metode Klasifikasi Naive Bayes dapat digunakan kembali untuk metode Decision Tree ini.